src/sdk/pynni/nni/compression/torch/lottery_ticket.py

```diff
@@ -0,0 +1,138 @@
+import copy
+import logging
+import torch
+import numpy as np
+from .compressor import Pruner
+
+_logger = logging.getLogger(__name__)
+
+
+class LotteryTicketPruner(Pruner):
+    """
+    This is a Pytorch implementation of the paper "The Lottery Ticket Hypothesis: Finding Sparse, Trainable Neural Networks",
+    following NNI model compression interface.
+
+    1. Randomly initialize a neural network f(x;theta_0) (where theta_0 follows D_{theta}).
+    2. Train the network for j iterations, arriving at parameters theta_j.
+    3. Prune p% of the parameters in theta_j, creating a mask m.
+    4. Reset the remaining parameters to their values in theta_0, creating the winning ticket f(x;m*theta_0).
+    5. Repeat step 2, 3, and 4.
+    """
+    def __init__(self, model, config_list, optimizer, reset_weights=True):
```

Comment:
maybe lr_scheduler also needs to be reset and it sometimes relies on the `epoch`, maybe init the `LotteryTicketPruner` with a `checkpoint` dict could be better, which contains the `state_dict` of the model, optimizer and scheduler, then prune the model in a loop, like
```
for _ in range(prune_iterations):
    for epoch in range(epoch_num):
        pruner.update_epoch(epoch)
        print('# Epoch {} #'.format(epoch))
        train(model, device, train_loader, optimizer)
```

Reply:
how to checkpoint lr_scheduler?

Reply:
good suggestion, refactored